### ISSUE TITLE
Signature has expired 

### ISSUE THREAD

**Issue by mnort9** [arcticleo/reso_api#7]
My oauth token seems to be expiring: ```JWT::ExpiredSignature (Signature has expired)```

I was able to successfully retrieve listings before the token expired. After looking into the source code, it should be handling the refresh automatically, right?

Also, I noticed the token is being written into the filesystem. This could be problematic in certain production environments like Heroku that have ephemeral filestystems. (issue above is locally, so not related)

**Comment by mnort9** [comment]
So it turns out my token is returned with no expiration time, but instead ```max_age_hours```. As a quick fix, I'm fetching a new one if there's an exception. 

```ruby
def oauth2_token
 payload = oauth2_payload
 token = JWT.decode(payload.token, nil, false) rescue nil # rescue if "Signature has expired"
 exp_timestamp = Hash(token.try(:first))["exp"].to_s
 expiration = DateTime.strptime(exp_timestamp, '%s').utc rescue DateTime.now.utc

 if token.present? && expiration > DateTime.now.utc # Fetch new token if unable to retrieve
  return payload.token
 else
  @oauth2_payload = oauth2_client.client_credentials.get_token
  File.write(oauth2_token_path, @oauth2_payload.to_hash.to_json)
  return @oauth2_payload.token
 end
end
```

**Comment by arcticleo** [comment]
Ah thanks for the observation. I had seen it too, but hadn't investigated the reason since it's incredibly rare in my usage. I have added handling of 401, 500, 503, and timeout errors in the perform_call method, but that's a great observation for the 401 and I will look at adding a proper fix for that in the oauth2_token method.

**Comment by arcticleo** [comment]
May I ask which software/system you're testing it against?

**Comment by mnort9** [comment]
Listhub

**Comment by mnort9** [comment]
Seems to be fixed in 0.4.8. Thanks!

**Comment by arcticleo** [comment]
Great! 👍 We've noticed a couple of other ListHub peculiarities as well, so stay tuned for more upcoming fixes/workarounds. 😊
